### ISSUE TITLE
Examine performance of individual data accessor methods of MemorySegmentIndexInput

### ISSUE THREAD

**Issue by ChrisHegarty** [apache/lucene#13325]
Examine performance of individual data accessor methods of MemorySegmentIndexInput, as compared to the ByteBufferIndexInput. Specifically getByte, getShort, and getBytes.

This issue has been filed to raise awareness of a seemingly perf degradation of MemorySegmentIndexInput when used heavily in decompression, see https://github.com/dacapobench/dacapobench/issues/264. I've not yet investigated the low-level segment guards, but maybe there is something here (or in the VarHandles) that is not optimising?

**Comment by mikemccand** [comment]
I've also opened https://github.com/mikemccand/luceneutil/issues/267 to understand why our nightly benchmarks didn't notice this.  @uschindler maybe you have an idea!

**Comment by uschindler** [comment]
The issue is analyzed already: The benchmark is wrong. It closes and opens indexes all the time. Due to the additional safety applied on closing indexes (to not crush the JVM with SIGSEGV), the JVM signals to all other threads that they need to go into safepoint to do the thread-local handshake.

The benchmakr should be rewritten to use a single IndexReader instance with many threads and not open a separate one for short time all the time in search threads. This causes thread-local-handshake havoc.

**Comment by uschindler** [comment]
So this issue is not in Lucene. The slowdown comes from incorrect use of Lucene. I am waiting for the confirmation by them.

**Comment by ChrisHegarty** [comment]
Thanks for your quick reply @uschindler. Your analysis makes sense to me.

**Comment by uschindler** [comment]
> Thanks for your quick reply @uschindler. Your analysis makes sense to me.

If they don't respond, I will rewrite their benchmark to only open the index once and then hammer it with different threads. If this is then identical speed the issue seen is confirmed to be caused by the thread-local handshakes.

Although this is incorrect use of Lucene, we should still try to investigate with Hotspot people why the handshakes/safepoints always lead to some kind of deoptimization in all threads accessing MemorySegments. This is not necessarily bad in normal workloads, but could lead to slowdowns in NRT use cases with high index refresh frequencies.

We could also add an option to MMapDirectory so you can make it ready for purely single-threaded use. I am not sure if this is useful in practise, but we could easily allow this. If another thread then reads from an index file it will trigger IllegalStateException.

**Comment by uschindler** [comment]
We have the confimation: https://github.com/dacapobench/dacapobench/issues/264#issuecomment-2084907716

I think we can close this issue. 

Anyways we should talk to the Hotspot people to figure out how we can improve the safepoint/thread-local handshake in a way that it does not get so expensive for other threads. To me it looks like on every close, all methods accessing MemorySegments get deoptimized, but I haven't looked into this.

**Comment by ChrisHegarty** [comment]
Thanks for jumping on this so quickly @uschindler.

**Comment by uschindler** [comment]
They identified that deoptimizations are indeed the cause of all this. When you trigger a thread-local handshake by closing an IndexInput, all code in other threads which runs at the same time and accessing memory segments gets deoptimized somehow. I have no idea how it enters optimized mode again, but this seems to cause the slowdown for that benchamrk.

I am not sure if we can open issues about this in OpenJDK to optimize this. Why does the code need to be deoptimized when we enter a safepoint/handshake? It just traps with a segv and enters safepoint.

**Comment by uschindler** [comment]
Hi,
I reopened this to make a further investigation.

@dsmiley  talked to me at Berlinbuzzwords and he also commented on #13146. He has seen a major slowdown on Apache Solr with Java 21 and using Lucene 9.8 (or like that). The issue there was partly 3x slower queries (as far as I understood) in an environment where many many Solr Cores with man many softcommits (IndexReader repoens for NRT) are happening at the same time. So think of about 1000 indexes on the same JVM and all of them are repopen indexes every second.

There is seems to be the problem that closing one IndexInput seems to deoptimize every other MemorySegment although not related / belonging to the same shared Arena.

I try to write a JMH benchmark without Lucene code to reproduce the issue and measure how large the issue is (depending on how often you close an IndexInput and how many parallel threads are using other MemorySegments, also those not tied to same Indexinput).

My idea is:
- Have many threads working on "hot" MemorySegments allocated with different shared Arenas.
- Have a few threads closing those Arenas.

Theoretically, the benchmark should not slow if the few threads close Arenas which are not bound to and hot workers.

What we have seen in dacapo bench is that it seems to affect all MemorySegments negatively, not only those which are affected by the close. of Arena.

Once I have the benchmark I will open issue at OpenJDK.

In the meantime, we could "preserve" the old Lucene 9.1 (the version without MemorySegments) in the "misc" module of Lucene as LegacyMMapDirectory. This would allow people to use it, but it is strinly discouraged. It won't work anymore with Java 24 (or around that time).

**Comment by magibney** [comment]
Practically speaking for Lucene, a lot of the issues we've seen with this seem to be around checksum `IndexInput` instances, which are opened and closed a lot at least under the SegmentReader ctor (as opposed to longer-lived MemorySegmentIndexInput that are used for searching, etc). This leads to a lot of closing for checksum `IndexInput` where with READ_ONCE, iiuc, there's little benefit to using MemorySegment -- or at least Arena.ofShared() -- at all.

In a toy program written to illustrate this issue (jdk only) I've been able to improve things somewhat (~6x, fwiw) by naively protecting access to `Arena.close()` by a singleton 1-permit semaphore. Not a solution by any means, but further indication that there are definitely issues in high-concurrency environments.

**Comment by mikemccand** [comment]
> improve things somewhat (~6x, fwiw) 

Uhm, ~6x seems a lot more than just a "somewhat" to me!  It's spooky that such a workaround (forcing only one thread to `.close()` at a time) can yield such gains in current JVMs...

Is this performance degradation serious enough that we should fix `FSDirectory.open`'s defaults to not use memory segments for `readonce` cases?

**Comment by uschindler** [comment]
The problem is that Lucene at moment needs shared arenas, as we do not know beforehand if it's is used by multiple threads.

In general the observation helps. So it looks like the thread local handshake causes issues with deoptinization if multiple threads cause a thread local handshake at same time 

Why not simply use the following workaround and measure it's effect:

Add a `static CLOSE_LOCK = new Object()` (it must be static) and synchronize the Arena.close() on it.

If this brings back performance, we can report this to jdk. At some point we can remove the lock when jdk version is greater xy.

**Comment by uschindler** [comment]
But yes. For read once we could use a confibed arena. In that case it will throw exception when accessed by multiple threads. In addition maybe IndexInput.clone() should throw exception and not allow clones.

**Comment by uschindler** [comment]
FYI, I had not much time to work on this. It would be great if we could open a bug report at OpenJDK and inform Maurizio about this issue.

I played a bit around with it - but that having a global lock on `Arena#close()` is helping is incredible! It may help the JDK devs to figure out what goes wrong!

Actually on all platforms, the thread-local handshake should not cause slowdown on "reader" threads. Only the actual `Arena#close()` call may be delayed or hang (actually not really different to a global lock as proposed). In no case it should cause a deoptimization on the other threads reading from MemorySegment!

To me this looks like an issue that theres a race inside the thread local handshakes so when multiple threads trigger this at same time, the JVM somehow gets into a state where it can only "escape" by deoptimizing all reader threads.

If we can create a simple benchmark without Lucene code showing this, we may help them to fix this quite soon! @magibney do you have some sample benchmark only using FileChannel#map on shared arenas and closing them with multiple threads? I was about to write something. If the performance of such benchmark restores when we synchronize, we have a proof whats going wrong.

I hope I can spend some time to work on this this week, but maybe @magibney has something already. I told @dsmiley on berlinbuzzrords that I will help!

**Comment by uschindler** [comment]
I sent an email to Maurizio, maybe he has an idea. This would help us to open a good issue. Maybe he already sees the problem. He has best overview on the internals when closing shared arenas.

**Comment by uschindler** [comment]
I also posted to panama-dev mailing list.

**Comment by magibney** [comment]
Haven't yet put together proper benchmarks, but I'm happy to share the toy program that I was using to evaluate the issue. I'm also not exactly sure that this is written to best target exactly the behavior that's problematic, but it seems to get pretty close, and with some tweaks can probably help guide in the right direction.

It's a [single java class](https://github.com/cowpaths/lucene/blob/michaelgibney/mmap-vs-arena-close/lucene/demo/src/java/org/apache/lucene/demo/ConcurrentMmapOrArenaClose.java), no dependencies.

Invoked as
```
java -cp [classpath_root] org.apache.lucene.demo.ConcurrentMmapOrArenaClose [bool_mmap] [n_threads] [n_idle_threads] [n_reps] [bool_semaphore]
```
where:
* [bool_mmap] if false uses Arena, if true uses legacy mmap
* [n_threads] number of threads concurrently opening and closing mmaps/shared arenas
* [n_idle_threads] number of threads that hold a region open but idle, closing only once all [n_threads] have completed
* [n_reps] the number of times that each of [n_threads] will open/close mmap/shared arena
* [bool_semaphore] for Arena, should access to Arena.close() be protected by a global single-permit semaphore?

A couple of things I've observed:
* [n_idle_threads] has a relatively small impact, but does have some impact
* [n_reps] is mostly to control the total amount of time to be long enough to be significant, but not so long that one gets tired of waiting for it to complete
* [bool_semaphore] seems to make a fairly consistent ~6x difference, even across various settings of [n_threads]
* the threads in [n_threads] are very busily opening/closing, so it's an artificial case of course. The concern is mostly that as [n_threads] varies it seems to scale very badly (esp. as compared to legacy/mmap)
* I did try Arena.ofConfined(); unsurprisingly it is very fast (but I didn't bother leaving that nor making that configurable because it's not really an option for Lucene atm).

**Comment by magibney** [comment]
> For read once we could use a confined arena

If this is an option I think it would likely have a heavy impact. Again, anecdotally from real-world experimentation in highly concurrent contexts, these READ_ONCE cases seemed to predominate in thread dumps. And the flip side of scaling badly with increased number of shared Arena.close() is that we should achieve an outsized impact from any reductions we can make (e.g., via READ_ONCE) to the number shared Arena.close() that are necessary!

>Add a static CLOSE_LOCK = new Object() (it must be static) and synchronize the Arena.close() on it.

Ah right, that's simpler. I initially used a semaphore because I was experimenting with varying the number of permits, but there was no benefit to even low levels of concurrency, so I quickly fell back to single-permit, not thinking that that's equivalent to just synchronizing on a global lock 😄

**Comment by uschindler** [comment]
Hi,

actually when looking at the code I don't understand what you are measuring. David said that reads are slow, but you actiually never read from the memory segment.

You only measure that it slows down closing, which is expected with shared arenas (as the thread-local handshake is needed). So It is actually expected that actually CLOSING takes longer, but what the others have observed is that also reads in totally unrelated threads get slower. So to me the test case here is not showing the full problem that the benchmark project and David Smileys issue.

So I think I need to write my own benchmark with JMH. Actually there should be 2 groups of threads:
- On thread randomly open and close memory segments. Thoe will definitely be slower!
- Another thread reading from another MemorySegment that was also mmapped but not closed. This one should stay fast. The others have seen slowdowns on reads there.

> > Add a static CLOSE_LOCK = new Object() (it must be static) and synchronize the Arena.close() on it.
> 
> Ah right, that's simpler. I initially used a semaphore because I was experimenting with varying the number of permits, but there was no benefit to even low levels of concurrency, so I quickly fell back to single-permit, not thinking that that's equivalent to just synchronizing on a global lock 😄 (EDIT: also, having read the jdk list email: the global lock doesn't make the issue go away entirely, but it seems to mitigate it by a constant factor -- scaling with concurrency is still far worse than linear though)

OK. But having a slowdown for Arena.close() is expected. What about other threads doing purely reads from MemorySegment?

Uwe

**Comment by uschindler** [comment]
Actually your benchmark has no warmup, I would expect it to be slow as the MemorySegment shared code is quite heavy.

**Comment by magibney** [comment]
Ah ok! I was indeed evaluating something different -- my concern was around closing per se, particularly the contribution of multiple concurrent closes to locking `IndexWriter`, and increased hard-commit times (apparently due to READ_ONCE checksum IndexInput)?

Even if I'm basically measuring "close throughput", the behavior still seems unusual.

Yeah, re: warmup ... this is a toy program, not a benchmark; I was sharing it just to be more transparent about what I was looking at. Possible mitigations/workarounds with READ_ONCE and synchronizing access to `Arena.close()` still may apply -- just something to investigate with real benchmarks. (A bit naively here, but fwiw: I'd expect the lack of warmup to if anything make the longer-running invocations skew to look _better_, whereas what we're seeing is the opposite).

I also have questions re: "having a slowdown for Arena.close() is expected". I was unaware of this, but I guess it's a question of degrees. How much worse is expected? All the numbers below with a grain of salt (no warmup measuring something that may not be directly as important), but just so there are some actual numbers here ...

Without global lock on `Arena.close()`, concurrency -> duration:
```
2: duration: 7ms
4: duration: 13ms
8: duration: 29ms
16: duration: 66ms
32: duration: 251ms
64: duration: 801ms
128: duration: 3585ms
256: duration: 12977ms
512: duration: 53724ms
1024: duration: 309576ms
```

with global lock on `Arena.close()`, concurrency -> duration:
```
2: duration: 10ms
4: duration: 18ms
8: duration: 30ms
16: duration: 52ms
32: duration: 112ms
64: duration: 267ms
128: duration: 749ms
256: duration: 2410ms
512: duration: 9180ms
1024: duration: 43172ms
```

**Comment by uschindler** [comment]
> I also have questions re: "having a slowdown for Arena.close() is expected". I was unaware of this, but I guess it's a question of degrees. How much worse is expected? All the numbers below with a grain of salt (no warmup measuring something that may not be directly as important), but just so there are some actual numbers here ...

The slower shared close is expected because that makes it different from the old code which could possibly crash the JVM. It will be slower and theres nothing we can do against it without having the risk to crash the JVM. The whole idea of the shared segments is:
- we don't want to slow down reads by always checking on each method invocation if the MemorySegment is closed. In multithreaded environments if one thread unmaps the ByteBuffer forcefully (thats what the usafe code is doing), another thread will segmentation fault, because the read form the ByteBuffer will hit an unmapped segment. This is because we cannot have an expensive check on every access.
- So we slow down the actual close by making it use some intrernal JVM magic: The close call will trigger an so called thread-local handshake. Basically it will signal to all threads that they should go into safepoint. Because all other threads cannot do this at the same time, the close() have to wait for all threads to do this. Depending on multithreaded load this may take some time. This ensures that before actually unmapping the memory segment, we know that all threads have "halted". This halting is done by some crazy magic using internal sigsegv at some specific points (e.g., when it exits a method or when it has a loop continuation.
- When all threads have halted, we can unmap the memory segment. And then all threads in the safepoint will make sure that they check the flag "was arena closed" before proceeding. This code will thor IllegalStateException("already closed").

Basically the thread local handshakes replace a volatile check on every read in every thread by some mechanism that the closing thread announces "i want to close arena, please all threads stop when you are working with memory segments". It then has to wait that all threads went into that state. Then it closes segment and then all threads can sping again after they have verified that the memory segment is still alive.

So basically we shift an expensive check per access to memory segment to a waiting time in close.

The issue that was observed is that obviously the "ask all other threads to go into thread local handshake" seems to have some effect on their performance.

But yes: When we do close() too often it gets expensive also for reading threads (and this is what others and David Smiley have seen).

The explanation here is a bit complicated and may be not fully correct (as theres a lot of magic involved). But basically the idea is to shift the expensive "check on every read from IndexInput" to "make close slower to allow all threads to finish work before we do close".

In Lucene with MappedByteBuffer we do neither (we just close the ByteBuffer's backing buffer and risk crashing the JVM). So the new code will be slower on highly concurrent environments. It depends on number of threads doing closes.

To your benchamrk: Without warmup this is in infact nonsense, as the whole magic is here happening in C2 compiler which requires a lot of warmup time. What you measure is a lot of interpreter logic which may cause loger waiting times on close() than with C2 where it should be highly optimized.

**Comment by uschindler** [comment]
In short: When you reduce the number of expensive close() calls on shared segments, reduce the number of times all threads need to go into thread-local handshakes.

Next to use confined segments for single-threaded reads (like checksumming), we may also think of combining multiple files that are always open at same time to use a single arena (because all files are opened/closed at same time).

I think what you see with the global lock is just that the number of thread-local handshakes goes down, because you close all memory segments in one go and not trigger multiple handshakes at same time.

I'd like to go the route suggested by @ChrisHegarty and use confined arenas for read-once. I think we can relatively simple implement this. Bugs will be found by our tests easily. In additon read-once segments should never be cloned.

But still we should create a real JMH benchmark that emulates a real use case: Many threds are reading from various segments and others closing a subset of them.

**Comment by magibney** [comment]
Thanks so much Uwe for the detailed explanation!

Confined arena sounds great for READ_ONCE -- in fact I was obliquely suggesting this also with my initial comment, but I wasn't sure whether there might be some reason this would not work for Lucene; glad to hear there's no such reason! And in fact I expect this may make an outsized contribution to the overall `Arena.close()` situation, based on what I've seen in stack traces (where threads seem to actually get stuck in `Arena.close()` seems mainly to be these one-off instances).

**Comment by uschindler** [comment]
Hi,
it looks like the underlying deoptimization issue is explained by Jorn Vernee: https://mail.openjdk.org/pipermail/panama-dev/2024-July/020532.html

> I've read the various github threads, and I think I have a good idea of 
> what's going on. Just to recap how shared arena closure works:
> 
> - We set the arena's isAlive = false (more or less)
> - We submit a handshake from the closing thread to all other threads
> - During that handshake, we check whether each thread is accessing the 
> arena we are trying to close.
>    - Unfortunately, this requires deoptimizing the top-most frame of a 
> thread, due to: JDK-8290892 [1]. But note that this is a 'one off' 
> unpacking of the top-most frame. After which the method finishes running 
> in the interpreter, and then goes back to executing compiled code.
> - If we find a thread accessing the same arena, we make it throw an 
> exception, to bail out of the access.
> - After the handshake finishes, we know that either: 1) no other thread 
> was accessing the arena, and they will see the up-to-date isAlive = 
> false because the handshake also works as a means of synchronization. Or 
> 2) if they were accessing the arena, they will get an exception.
> - After that it's safe to free the memory.
> 
> So, if shared arenas are closed very frequently, as seems to be the case 
> in the problematic scenarios, we're likely overwhelming all other 
> threads with handshakes and deoptimizations.
> 
> Addressing JDK-8290892 would mean that we only need to deoptimize 
> threads that are actually accessing the arena that is being closed. 
> Another possible interim improvement could be to only deoptimize threads 
> that are actually in the middle of accessing a memory segment, not just 
> all threads. That's a relatively low-hanging fruit I noticed recently, 
> but haven't had time to look into yet. I've filed: 
> https://bugs.openjdk.org/browse/JDK-8335480
> 
> This would of course still leave the handshakes, which also require a 
> bunch of VM-internal synchronization between threads. Shared arenas are 
> really only meant for long-lived lifetimes. So, either way, you may want 
> to consider ways of grouping resources into the same shared arena, to 
> reduce the total number of closures needed, or giving users the option 
> of indicating that they only need to access a resource from a single 
> thread, and then switching to a confined arena behind the scenes (which 
> is much cheaper to close, as you've found).
> 
> Jorn
> 
> [1]: https://bugs.openjdk.org/browse/JDK-8290892

**Comment by uschindler** [comment]
Here is some response I sent: https://mail.openjdk.org/pipermail/panama-dev/2024-July/020535.html

Actually the ideas I have:
- Use READ_ONCE in combination with single threaded confined Arenas
- Possibly group multiple files into one arena. One Idea that came into my mind: MMapDirectory groups files belonging to the same segment together and uses a single Arena for them. Behind the senes it keeps a refcount for each group of files. So whenever a file is opened the refcount for that group increases and when it is closed it decreases. When the refcount goes to 0 the Arena is closed.

**Comment by ChrisHegarty** [comment]
> Here is some response I sent: https://mail.openjdk.org/pipermail/panama-dev/2024-July/020535.html

Thanks @uschindler - this is very helpful.

> Actually the ideas I have:
> 
> * Use READ_ONCE in combination with single threaded confined Arenas. Also disallow cloning of such IndexInputs (because thats mainly done for multi-threaded access).

A quick PR to see what breaks with this - #13535.  Good new is that nothing seems to break.

> * Possibly group multiple files into one arena. One Idea that came into my mind: MMapDirectory groups files belonging to the same segment together and uses a single Arena for them. Behind the scenes it keeps a refcount for each group of files. So whenever a file is opened the refcount for that group increases and when it is closed it decreases. When the refcount goes to 0 the Arena is closed.

Seems reasonable. Will have to benchmark to see how much it will improve things in practice.

**Comment by magibney** [comment]
> Possibly group multiple files into one arena. One Idea that came into my mind: MMapDirectory groups files belonging to the same segment together and uses a single Arena for them. Behind the scenes it keeps a refcount for each group of files. So whenever a file is opened the refcount for that group increases and when it is closed it decreases. When the refcount goes to 0 the Arena is closed.

This seems really promising as well. I may try to pick this up unless anyone else is already looking into it? This seems like the right thing to do semantically even, in addition to the potential performance benefits.

**Comment by uschindler** [comment]
> > Possibly group multiple files into one arena. One Idea that came into my mind: MMapDirectory groups files belonging to the same segment together and uses a single Arena for them. Behind the scenes it keeps a refcount for each group of files. So whenever a file is opened the refcount for that group increases and when it is closed it decreases. When the refcount goes to 0 the Arena is closed.
> 
> This seems really promising as well. I may try to pick this up unless anyone else is already looking into it? This seems like the right thing to do semantically even, in addition to the potential performance benefits.

I had no time to look into this. It should be doable with a instance ConcurrentHashMap, not sure where to place this instance (as it is also using those providers). So the problematic part is where to place this (as it cannot be in MMapDirectory where it should belong as it uses Java 21 preview apis).

You can give it a try, but i have the feeling it gets tricky on the actual implementation. We may need to add some abstraction into MMapDirectory instances (the ConcurrentHashMap) with the refcounter and a pointer to some "hidden" payload (the areana) which is picked by the private MemorySegmentIndexInputProvider.

We can work together. If you have a better idea, tell me.

**Comment by magibney** [comment]
I extended the `Arena` interface to track refCount for https://github.com/apache/lucene/pull/13555 -- curious what you think of this approach?

**Comment by uschindler** [comment]
We also have another PR doing the same than @magibney's: https://github.com/apache/lucene/pull/13570

We also got some updates on the JDK front. Next to the fixes here (use confined segments for READONCE and grouping shared arenas) there are a few suggestions:

I have the feeling that Elasticsearch does not see the issue like Solr does under high load because the concurrency is different. @ChrisHegarty Is it true that Elasticsearch refreshes indexes in a sepate thread, so all close() calls are from a few threads only? In large Solr the installations the problem is mainly that every core looks like having its own threads that refresh indexes. Maybe it would be a good idea in Apache Solr to somehow have a maintenance thread which is responsible for refreshing indexes. The original patch by @magibney already showed that a static singleton lock around `Arena#close` looks like it helps. This in fact makes all closes sequential.

@dsmiley Maybe that's another solution to prevent this issue from happening: Introduce a global "maintenance task" in Solr which is responsible to reopen indexes. We could include that also in the SearcherManager tool? (cc @mikemccand).

I think with all those fixes here we have fixed most of the overhead.

**Comment by uschindler** [comment]
Yeah: https://github.com/openjdk/jdk/pull/20158#issuecomment-2228916752

**Comment by dsmiley** [comment]
What does "refreshing indexes" mean exactly?  Solr does not have special threads for opening/closing indexes; they will typically happen on the common Jetty request thread pool.  If the suggestion is _intended_ to ensure that only one thread open/close an index, I don't see why that would be a "maintenance task" but I do see it could easily be done simply by ensuring only one thread does this work on-demand.  But I'd be concerned indexes would open slowly if you have many.  Any way; feels like a work-around to the JDK bug and not something desirable in its own right.

**Comment by uschindler** [comment]
It is not a JDK bug, it is expected behaviour.

**Comment by uschindler** [comment]
Basically the idea is to close indexes asynchronous in a single thread doing this to not block main threads.

**Comment by uschindler** [comment]
To be clear: With raising number of threads, the process of closing an indexinput gets slower. With recent Lucene improve,ments this is mitigated a bit by grouping files together and read READONCE files with a confined arena, but for all other IndexInputs which are in multithreaded use, the closing thread has to wait for the handshake on ALL other threads, so it does not scale well when there are many threads. This is needed for safety to ensure no other thread is accesssing the index while we are closing. There is no way to work around that as it is the required safety applied by the JVM (that we are currently missing in Lucene, this is why you can crush the whole VM when you close an IndexReader while threads are accessing it).

Because of this, the JDK maintainers recommend to possibly do the closing of indexes asynchronously in production environments. So basically in the normal processing of Solr, reopen the index as usual, but when actually closing the old IndexReader defer this to a single-threaded helper. This might also be a good idea for Elasticsearch.

This would allow to open the fresh indexes as soon as possible, but the old segments get closed in a delayed way. We can't implement this inside Lucene as we have no access to thread pools, it is up to the user to possibly close indexes async.

**Comment by dsmiley** [comment]
Solr already takes care to ensure a SolrCore is closed while nobody is accessing it via its refcount mechanism on the SolrCore.  I understand the JVM doesn't know that, and needs to take precautions.  That it's O(threads) means the Solr project should prioritize taming its 10k-bounded Jetty thread pool for this new reason.  

> basically reopening IndexReaders (closing old segments, open new segments). Basically a Solr softcommit, but also applies to commits

Woah; per commit! I thought you just meant the equivalent of opening and closing a SolrCore, which has an index -- somewhat rare.  Are you saying that, ideally, Solr should close the _previous_ IndexReader (tied to a Searcher that will close) asynchronously?  And you suggested by one thread... does it need to be one thread for the whole JVM (this worries me) or one per core?

> the JDK maintainers recommend to possibly do the closing of indexes asynchronously in production environments

It's okay if closing a SolrCore is a bit slower.  In practice, this means a core UNLOAD request would take longer thus deleting a collection takes longer (albeit is already parallelized per replica).  Shutting down Solr would take longer; okay.

**Comment by uschindler** [comment]
Hi David,
sorry for late response. In fact there problem is quite simple. Every IndexInput#close() costs time in the thread calling this. This time is linear to the number of threads (in plain Java 19-21 at moment).
With the fixes by JDK people it will change a bit:
- It will still be somehow linear to the total number of threads, but gets faster when not all threads are actually inside a MemorySement read
- The will be no overhead anymore on consuming threads that don't use MemorySegments; at the moment it deoptimizes all threads.

So with the JDK changes to be applied to Java 22+ (and maybe also 21) the overhead will be almost zero for threads querying Lucene. The work for closing only minimally changes (if at all).

So the best we can do is: Don't concurently close IndexInputs in the main threads because on heavy loaded Solr clusters this can take a lot of time. So the idea is to do all "maintenance" on Lucene IndexSearchers/IndexReaders in a small & separate pool of threads. Actively reopen a DirectoryReader (after commit or softcommit) can be done in main thread, but closing the old one can easily be deferred to a small maintenance thread pool.

There is no way around this, as the ByteBuffer based MMap will no lobger be possible with later Java versions, as sun.misc.Unsafe will be decommissioned in Java 24 latest. So the new code and its additional safety we have to live with. And that's good. My suggestion would be to to also rethink Solr's thread usage. 10.000 threads is incredibly wrong! Possibly use virtual threads. If you do this, the problem will also go away.

Uwe

**Comment by uschindler** [comment]
P.S.: The code optimizations were committed for JDK 23 already. Backport is under discussion.

**Comment by dsmiley** [comment]
Thanks for your detailed response!

> The will be no overhead anymore on unrelated threads that don't use MemorySegments; at the moment it deoptimizes all threads.

Much of the fundamental tasks of a search engine access the index and those will use the MemorySegment API.  Won't the "deoptimize" part still be a problem?

**Comment by uschindler** [comment]
> Thanks for your detailed response!
> 
> > The will be no overhead anymore on unrelated threads that don't use MemorySegments; at the moment it deoptimizes all threads.
> 
> Much of the fundamental tasks of a search engine access the index and those will use the MemorySegment API. Won't the "deoptimize" part still be a problem?

No, see benchmarks. In fact only the top frame is deoptimized, but once the processing can go on the top frame is replaced by the orptimized code again. This is a typical VM operation that happens under other circumstances, too.

**Comment by dsmiley** [comment]
I set a breakpoint in a debugger on IndexInput.close while running a Solr test and it's called in **lots** of places separately from closing an IndexReader.  Lucene FieldsIndexWriter.finish *reads* a file it just finished writing, for example.  If the underlying problem is more closely associated to the IndexInput and not the IndexReader, shouldn't IndexInput.close be internally async, and thus be a Lucene level change?

I also set a breakpoint on IndexReader.close.  Lucene's SpellChecker.swapSearcher is invoked when the searcher changes (e.g. from a commit) via Solr's SpellCheckerListener.  Hopefully the only effect leaving this is commits being slower.

**Comment by uschindler** [comment]
> I set a breakpoint in a debugger on IndexInput.close while running a Solr test and it's called in **lots** of places separately from closing an IndexReader. Lucene FieldsIndexWriter.finish _reads_ a file it just finished writing, for example.

This ChecksumIndexInput should be opened with READONCE, so it is confined. So there's no cost on closing:

https://github.com/apache/lucene/blob/cc3b412183c7acbb4a6c498e9babffba4e7d1f28/lucene/core/src/java/org/apache/lucene/codecs/lucene90/compressing/FieldsIndexWriter.java#L123

Uwe

**Comment by uschindler** [comment]
All ChecksumIndexInput are automatically opened using READONCE in main branch as those cannot get consumed by multiple threads. In the backport it's manually done; if we missed one place please open issue.

**Comment by dsmiley** [comment]
I filed an issue in Solr: https://issues.apache.org/jira/browse/SOLR-17375

**Comment by Bukhtawar** [comment]
Noticing a related issue in OpenSearch https://github.com/opensearch-project/OpenSearch/issues/16180
